File: yumi/edge/template/yumi_tools/kotlin/src/main/kotlin/io/yumi/edge/YumiSetup.kt
package io.yumi.edge

import io.yumi.sdk.AgentOptions
import io.yumi.sdk.YumiAgent
import io.yumi.sdk.RegisterOptions
import io.yumi.sdk.ToolHandler
import io.yumi.sdk.ToolParameter

fun initYumi() {
    val agent = YumiAgent(
        AgentOptions(
            connectionCode = null,
            edgeName = "My Kotlin App",
            envPath = null,
        ),
    )
    agent.register(
        RegisterOptions(
            name = "hello",
            description = "Say hello to someone",
            parameters = listOf(
                ToolParameter(
                    name = "name",
                    typeName = "string",
                    description = "Person to greet",
                ),
            ),
            handler = ToolHandler { args -> "Hello, ${args.string("name")}!" },
        ),
    )
    // Example tool — replace with your own. Registered in "pinned" exposure mode
    // so it is always exposed to the agent.
    agent.register(
        RegisterOptions(
            name = "ping",
            description = "Ping the edge and echo a message back",
            parameters = listOf(
                ToolParameter(
                    name = "message",
                    typeName = "string",
                    description = "Text to echo back.",
                    isRequired = false,
                ),
            ),
            mode = "pinned",
            handler = ToolHandler { args -> "pong: ${args.string("message")}" },
        ),
    )
    agent.runInBackground()
}

File: yumi/edge/template/yumi_tools/csharp/YumiSetup.cs
using Yumi;

/// <summary>
/// Yumi edge setup — register your tools here.
/// Call <see cref="YumiSetup.InitYumi"/> from your application's entry point.
/// </summary>
public static class YumiSetup
{
    // ── Configuration ──
    // Set your connection code here (LAN: "yumi-lan_...", remote: "yumi_...").
    // Leave empty to read from YUMI_CONNECTION_CODE env var or yumi_tools/.env.
    private const string YumiConnectionCode = "";
    private const string YumiEdgeName = "My C# App";

    public static YumiAgent InitYumi()
    {
        var agent = new YumiAgent(YumiConnectionCode, YumiEdgeName);

        // Register your tools below.

        // Example tool — replace with your own. Echoes a message back so you can
        // confirm the edge is connected. Pinned so it's always exposed to the agent.
        agent.Register(new RegisterOptions()
            .SetName("ping")
            .SetDescription("Ping the edge and echo a message back")
            .SetMode("pinned")
            .SetParameters(
                new ToolParameter("message", "string", "Text to echo back", required: false)
            )
            .SetHandler(args =>
            {
                var message = args.GetString("message", "hello");
                return $"pong: {message}";
            })
        );

        agent.Register(new RegisterOptions()
            .SetName("hello")
            .SetDescription("Say hello to someone")
            .SetParameters(
                new ToolParameter("name", "string", "Person to greet")
            )
            .SetHandler(args =>
            {
                var name = args.GetString("name", "World");
                return $"Hello, {name}!";
            })
        );

        // Example: tool with confirmation required
        // agent.Register(new RegisterOptions()
        //     .SetName("dangerous_action")
        //     .SetDescription("Do something irreversible")
        //     .SetRequireConfirmation(true)
        //     .SetHandler(args => "done")
        // );

        agent.RunInBackground();
        return agent;
    }
}
